158: fix std::vector serialization

cz4rs opened the pull request:
fixes #158 
fixes #159 

- do not traverse spare memory past the last element of vector (but account for it when footprinting)
- use data() to guarantee valid pointer to underlying array (even for empty container)
- avoid dereferencing null pointers on dispatch
- add adequate unit test

cz4rs committed:
#158: container: fix std::vector serialization

- do not traverse spare memory past the last element of vector
- use data() to guarantee valid pointer to underlying array (even for empty
container)
- avoid dereferencing null pointers
- add adequate unit test

cz4rs commented on src/checkpoint/dispatch/dispatch.impl.h:
```diff
@@ -122,7 +122,9 @@ T* Standard::construct(SerialByteType* mem) {
 template <typename Serializer, typename T>
 inline void serializeArray(Serializer& s, T* array, SerialSizeType const len) {
-  Traverse::with<T, Serializer>(*array, s, len);
+  if (array) {
```
I'm not sure if this is enough:

> https://en.cppreference.com/w/cpp/container/vector/data
If size() is 0, data() may or may not return a null pointer.

PhilMiller replied:
How about `if (len > 0)`?

cz4rs replied:
ad53ede crashed with non-null pointer (size = 0, capacity > 0), but `if (len > 0)` seems to work

PhilMiller commented on tests/unit/test_footprinter.cc:
```diff
@@ -474,48 +515,15 @@ TEST_F(TestFootprinter, test_vector) {
-    s | shared_pointer;
+  {
+    std::vector<TestDerived2> v;
+    EXPECT_EQ(
```
Could this potentially return greater, if the default-constructed vector has a non-zero capacity?

cz4rs replied:
~I was just testing that case (ad53ede) - it was too early to celebrate victory~

cz4rs replied:
yes, this is implementation dependent, so I got lucky - changed to `GE`

cz4rs committed:
#158: container: do not traverse empty arrays

- check for length before traversal, as empty container might have
reserved capacity (resulting in non-null pointer)
- test: do not assume capacity for empty std::vector

PhilMiller commented on tests/unit/test_footprinter.cc:
```diff
@@ -474,48 +515,21 @@ TEST_F(TestFootprinter, test_vector) {
+    EXPECT_GE(
+      checkpoint::getMemoryFootprint(v),
+      sizeof(v)
+    );
```
I think this test could still fail spuriously on some platforms, if a default-constructed `vector<T> v` has `v.capacity() > 0`.

cz4rs replied:
`EXPECT_GE` (greater or equal) should cover this

PhilMiller replied:
Oops, sorry, I misread that this says `EXPECT_GE` now.

PhilMiller commented on tests/unit/test_footprinter.cc:
```diff
@@ -474,48 +515,21 @@ TEST_F(TestFootprinter, test_vector) {
-    s | raw_pointer;
-    s | shared_pointer;
+    v.reserve(3);
+    EXPECT_EQ(
```
I think this case also needs to be `EXPECT_GE` though - there's no requirement that the memory allocated by `reserve` be a 'tight fit'

cz4rs replied:
you beat me by 90 seconds with this comment :)

cz4rs replied:
in the end, it will always work because the comparison uses `v.capacity()` (only `reserve` uses magic number)

PhilMiller replied:
I see, great.

cz4rs commented on tests/unit/test_footprinter.cc:
```diff
@@ -474,48 +515,21 @@ TEST_F(TestFootprinter, test_vector) {
+    v.reserve(3);
+    EXPECT_EQ(
+      checkpoint::getMemoryFootprint(v),
+      sizeof(v) + v.capacity() * sizeof(TestDerived2)
```
~ok, to avoid surprises, this could also fail on some platforms (if capacity was bigger than 3 by default):~
> reserve() cannot be used to reduce the capacity of the container; to that end shrink_to_fit() is provided. 

~`shrink_to_fit()` incoming~

PhilMiller replied:
It's not just if capacity is greater than 3 by default, it's also if the `reserve` call gives back a capacity greater than 3. This could happen if it rounds to some unit that it knows the allocator works in, for instance, or if the library implements http://wg21.link/p0401

lifflander reviewed (approved):
This look solid to me
